dizoo/classic_control/cartpole/entry/cartpole_dqn_gail_main.py

```diff
@@ -0,0 +1,162 @@
+    learner.call_hook('before_run')
+
+    # Accumulate plenty of data at the beginning of training.
+    if cfg.policy.get('random_collect_size', 0) > 0:
+        action_space = collector_env.env_info().act_space
+        random_policy = PolicyFactory.get_random_policy(policy.collect_mode, action_space=action_space)
+        collector.reset_policy(random_policy)
+        collect_kwargs = commander.step()
+        new_data = collector.collect(n_sample=cfg.policy.random_collect_size, policy_kwargs=collect_kwargs)
+        replay_buffer.push(new_data, cur_collector_envstep=0)
+        collector.reset_policy(policy.collect_mode)
+    for _ in range(max_iterations):
+        collect_kwargs = commander.step()
+        # Evaluate policy performance
+        if evaluator.should_eval(learner.train_iter):
+            stop, reward = evaluator.eval(learner.save_checkpoint, learner.train_iter, collector.envstep)
+            if stop:
+                break
+        new_data_count, target_new_data_count = 0, cfg.reward_model.get('target_new_data_count', 1)
+        while new_data_count < target_new_data_count:
+            new_data = collector.collect(train_iter=learner.train_iter, policy_kwargs=collect_kwargs)
+            new_data_count += len(new_data)
+            # collect data for reward_model training
+            reward_model.collect_data(new_data)
+            replay_buffer.push(new_data, cur_collector_envstep=collector.envstep)
+        # update reward_model
+        reward_model.train()
+        reward_model.clear_data()
+        # Learn policy from collected data
+        for i in range(cfg.policy.learn.update_per_collect):
+            # Learner will train ``update_per_collect`` times in one iteration.
+            train_data = replay_buffer.sample(learner.policy.get_attribute('batch_size'), learner.train_iter)
+            if train_data is None:
+                # It is possible that replay buffer's data count is too few to train ``update_per_collect`` times
+                logging.warning(
+                    "Replay buffer's data can only train for {} steps. ".format(i) +
+                    "You can modify data collect config, e.g. increasing n_sample, n_episode."
+                )
+                break
+            # update train_data reward
+            reward_model.estimate(train_data)
+            learner.train(train_data, collector.envstep)
+            if learner.policy.get_attribute('priority'):
+                replay_buffer.update(learner.priority_info)
+        if cfg.policy.on_policy:
+            # On-policy algorithm must clear the replay buffer.
+            replay_buffer.clear()
+
+    # Learner's after_run hook.
+    learner.call_hook('after_run')
+    # save reward model
+    path = os.path.join(cfg.exp_name, 'reward_model', 'ckpt')
+    if not os.path.exists(path):
+        try:
+            os.makedirs(path)
+        except FileExistsError:
+            pass
+    path = os.path.join(path, 'ckpt_last.pth.tar')
```

Comment:
maybe 'ckpt_final_reward_model.pth.tar' is more clearer? ( as we have ckpt_best.pth.tar before, referring to the best model for the training the policy.)

Reply:
Line 143: `path = os.path.join(cfg.exp_name, 'reward_model', 'ckpt')` creates a new path (`reward_model/ckpt`) where to save the state dict of the reward model. Given the explicit name, it should be clear that are pth.tar files inside that directory refer to the state dict of the reward model.